src/HttpClient/AuthenticatedHttpClient.php

```diff
@@ -55,36 +51,62 @@ public function __construct(
      */
     public function sendRequest($httpMethod, $uri, array $headers = [], $body = null)
     {
-        if (null === $this->accessToken) {
+        if (null === $this->authentication->getAccessToken()) {
             $tokens = $this->authenticationApi->authenticateByPassword(
                 $this->authentication->getClientId(),
                 $this->authentication->getSecret(),
                 $this->authentication->getUsername(),
                 $this->authentication->getPassword()
             );
 
-            $this->accessToken = $tokens['access_token'];
-            $this->refreshToken = $tokens['refresh_token'];
+            $this->authentication
+                ->setAccessToken($tokens['access_token'])
+                ->setRefreshToken($tokens['refresh_token']);
         }
 
         try {
-            $headers['Authorization'] =  sprintf('Bearer %s', $this->accessToken);
+            $headers['Authorization'] =  sprintf('Bearer %s', $this->authentication->getAccessToken());
             $response = $this->basicHttpClient->sendRequest($httpMethod, $uri, $headers, $body);
         } catch (UnauthorizedHttpException $e) {
+            $response = $this->resendUnauthorizedRequest($httpMethod, $uri, $headers, $body);
```

Comment:
it seems really weird when I read the code and we should talk it with @skeleton. 
But at least, explain why you do it in the phpdoc of the function.